src/libaktualizr/storage/sqlstorage.cc

```diff
@@ -12,6 +12,8 @@
 #include "sql_utils.h"
 #include "utilities/utils.h"
 
+std::mutex SQLite3Guard::m_;
```

Comment:
I'm a bit confused here. This treats this mutex like a global instead of as a member variable. Is that really what we have to do? If so, then can you add a comment that explains why it has to be that way?

Reply:
If the mutex is a member of SQLite3Guard, it will be destroyed with the SQLite3Guard variable. And currently the way of our code using SQLite3Guard is that define it in the stack of functions which will access database, so it will be destroyed when the function returned. IMO, it means every function using a new mutex, so it will not really lock. Please correct me if I made something wrong.

Reply:
> currently the way of our code using SQLite3Guard is that define it in the stack of functions which will access database, so it will be destroyed when the function returned. IMO, it means every function using a new mutex, so it will not really lock.

Isn't that what we want? We should only lock the database when we are writing to it, and then immediately release. Or do we really want to hold on to the mutex for the duration of the Storage object? I don't really see what the point of that is if the Storage is basically a singleton.

Also, I just noticed that the `extern std::mutex sql_mutex` declared in sql_utils.h appears to be unused and an obsolete relic of previous rounds of locking logic.

Reply:
Sorry, just realized what you meant now. Yes, the mutex does have to live a higher level to be meaningful. But wouldn't a member variable of SqlStorage still be preferable to this?

Reply:
If it's a member of SqlStorage, I think I will have two options to implement the lock. The first one is that add a lock_guard to every function of SqlStorage, which will lock/unlock this mutex, I don't think it's more graceful than using SQLite3Guard, which is already used in every function. The second option is let the mutex be member of SqlStorage, but still use SQLite3Guard to lock it. I think it's also not graceful. So I choose to use a static member of SQLite3Guard.

Reply:
This was mentioned on Tuesday I think.
It's also my opinion that we'd better have the lock live in some instance instead of being static if we can. Obvious drawback is that we'd have separate instances of Storage locking each others out, even if they pertain to different files.

The SQLite3Guard can be associated to a storage-specific lock in `SQLStorageBase::dbConnection()`.

Reply:
Ah, SQLStorageBase::dbConnection() is a good place to do it for most cases, but there is still some problems in class SQLTargetWHandle. If I keep a pointer/reference of the mutex of Storage in this WHandle class, it will be a little weird (and maybe dangerous). Actually I'm a little confused why we need the class SQLTargetWHandle, IMHO, it's function might be combine into Storage.

Reply:
Oh, it's also possible to keep a pointer of storage in that WHandle class, I think it will be better than keep a pointer of mutex, I will try it.

Reply:
Another problem is that now we could init SQLite3Guard with the path of database, that means we didn't associate a Storage object to it. If we want the mutex be a member of Storage, maybe we'd better avoid init it in this way.

Reply:
Keeping a pointer to storage in WHandle seems fine.

And you want to pass a mutex in `SQLite3Guard`'s constructor, is that right?
As long as the other calls to the constructors are in existing tests which don't try to access the db in parallel with SQLStorage, that's maybe fine if the lock is made optional.

We can discuss again on Friday to clarify the question.

Reply:
Yes, I want to pass a mutex into SQLite3Guard's constructor, and use a shared_ptr of mutex to keep it. But when I start coding, I found I still feel not very good for this design, it still make a pointer to a member variable of another class.
And yes, I also think make the lock be optional will make the existing test code work.

Reply:
Looks it's not a problem to use a pointer of mutex SQLite3Guard, because we usually only use SQLite3Guard as local variable. And I haven't found another way which is graceful enough and will not need to restruct all of the storage code. So I think I will still use this solution.